src/ccl_cls.c

```diff
@@ -0,0 +1,1336 @@
+      tkarr[ik]=transfer_wrap(il,lkarr[ik],cosmo,w,clt,status);
+    if(*status) {
+      free(clt->n_k);
+      free(lkarr);
+      free(tkarr);
+      break;
+    }
+
+    //Initialize spline for this ell
+    clt->spl_transfer[il]=ccl_spline_init(nk,lkarr,tkarr,0,0);
+    if(clt->spl_transfer[il]==NULL) {
+      free(lkarr);
+      free(tkarr);
+      *status=CCL_ERROR_MEMORY;
+      strcpy(cosmo->status_message,"ccl_cls.c: compute_transfer(): memory allocation\n");
+      break;
+    }
+    free(lkarr);
+    free(tkarr);
+  }
+  if(*status) {
+    free(clt->n_k);
+    int ill;
+    for(ill=0;ill<il;ill++)
+      ccl_spline_free(clt->spl_transfer[il]);
+    free(clt->spl_transfer);
+  }
+
+  clt->computed_transfer=1;
+}
+
+static double transfer(int il,double lk,ccl_cosmology *cosmo,
+		       CCL_ClWorkspace *w,CCL_ClTracer *clt,int *status)
+{
+  if(il<w->l_limber) {
+    if(!(clt->computed_transfer))
+      compute_transfer(clt,cosmo,w,status);
+
+    return ccl_spline_eval(lk,clt->spl_transfer[il]);
+  } else {
+    return transfer_wrap(il,lk,cosmo,w,clt,status);
+  }
+}
+
+//Params for power spectrum integrand
+typedef struct {
+  int il;
+  ccl_cosmology *cosmo;
+  CCL_ClWorkspace *w;
+  CCL_ClTracer *clt1;
+  CCL_ClTracer *clt2;
+  int *status;
+} IntClPar;
+
+//Integrand for integral power spectrum
+static double cl_integrand(double lk,void *params)
+{
+  double d1,d2;
+  IntClPar *p=(IntClPar *)params;
+  d1=transfer(p->il,lk,p->cosmo,p->w,p->clt1,p->status);
+  d2=transfer(p->il,lk,p->cosmo,p->w,p->clt2,p->status);
+
+  return pow(10.,3*lk)*d1*d2;
+}
+
+//Figure out k intervals where the Limber kernel has support
+//clt1 -> tracer #1
+//clt2 -> tracer #2
+//l    -> angular multipole
+//lkmin, lkmax -> log10 of the range of scales where the transfer functions have support
+static void get_k_interval(ccl_cosmology *cosmo,CCL_ClWorkspace *w,
+			   CCL_ClTracer *clt1,CCL_ClTracer *clt2,int l,
+			   double *lkmin,double *lkmax)
+{
+  double chimin,chimax;
+  int cut_low_1=0,cut_low_2=0;
+
+  //Define a minimum distance only if no lensing is needed
+  if((clt1->tracer_type==CL_TRACER_NC) && (clt1->has_magnification==0)) cut_low_1=1;
+  if((clt2->tracer_type==CL_TRACER_NC) && (clt2->has_magnification==0)) cut_low_2=1;
+
+  if(l<w->l_limber) {
+    chimin=2*(l+0.5)/ccl_splines->K_MAX;
+    chimax=0.5*(l+0.5)/ccl_splines->K_MIN_DEFAULT;
+  }
+  else {
+    if(cut_low_1) {
+      if(cut_low_2) {
+	chimin=fmax(clt1->chimin,clt2->chimin);
+	chimax=fmin(clt1->chimax,clt2->chimax);
+      }
+      else {
+	chimin=clt1->chimin;
+	chimax=clt1->chimax;
+      }
+    }
+    else if(cut_low_2) {
+      chimin=clt2->chimin;
+      chimax=clt2->chimax;
+    }
+    else {
+      chimin=0.5*(l+0.5)/ccl_splines->K_MAX;
+      chimax=2*(l+0.5)/ccl_splines->K_MIN_DEFAULT;
+    }
+  }
+
+  if(chimin<=0)
+    chimin=0.5*(l+0.5)/ccl_splines->K_MAX;
+
+  *lkmax=fmin( 2,log10(2  *(l+0.5)/chimin));
+  *lkmin=fmax(-4,log10(0.5*(l+0.5)/chimax));
```

Comment:
For the non-Limber integrator, we indeed integrate between K_MIN and K_MAX.

For the Limber integrator, however, we speed up the calculation by using only the interval of k-s where the integrand has non-negligible support. Basically, because chi~ell/k, and for e.g. number counts n(z) only has support on a limited range of chi, the integrand in the Limber approximation only has support over a limited range of k. Using this limit is useful for 2 reasons:
1- It can speed up the integration
2- It actually avoids the GSL integrator failing. If the integrand is very compact and you give the integrator a very broad integration range, it often returns zero after evaluating the integrand in a sufficiently large number of points that miss the actual range where the integrand is non-zero.

I've documented this in #506